exporter/prometheusremotewriteexporter/exporter.go

```diff
@@ -148,69 +150,115 @@ func validateAndSanitizeExternalLabels(cfg *Config) (map[string]string, error) {
+		sb.WriteByte(',')
+	}
+	return sb.String()
+}
+
+// handleExport partitions the time series map into N arrays and creates N workers to process them.
 func (prwe *prwExporter) handleExport(ctx context.Context, tsMap map[string]*prompb.TimeSeries) error {
 	// There are no metrics to export, so return.
 	if len(tsMap) == 0 {
 		return nil
 	}
 
-	// Calls the helper function to convert and batch the TsMap to the desired format
-	requests, err := batchTimeSeries(tsMap, maxBatchByteSize)
-	if err != nil {
-		return err
-	}
-	if !prwe.walEnabled() {
-		// Perform a direct export otherwise.
-		return prwe.export(ctx, requests)
-	}
+	// Determine the number of partitions (N) based on the concurrency limit and the number of requests
+	concurrencyLimit := int(math.Min(float64(prwe.concurrency), float64(len(tsMap))))
 
-	// Otherwise the WAL is enabled, and just persist the requests to the WAL
-	// and they'll be exported in another goroutine to the RemoteWrite endpoint.
-	if err = prwe.wal.persistToWAL(requests); err != nil {
-		return consumererror.NewPermanent(err)
+	// Partition the time series map into N arrays
+	partitions := partitionTimeSeries(tsMap, concurrencyLimit)
+
+	// Create N workers
+	wg := sync.WaitGroup{}
+	wg.Add(concurrencyLimit)
+
+	for i := 0; i < concurrencyLimit; i++ {
+		// Submit one array per worker
+		go func(partition map[string]*prompb.TimeSeries) {
+			defer wg.Done()
+
+			requests, err := batchTimeSeries(partition, maxBatchByteSize)
+			if err != nil {
+				err = consumererror.NewPermanent(err)
+				prwe.settings.Logger.Error("Failed to batch time series:", zap.Error(err))
+			}
+
+			if !prwe.walEnabled() {
+				// WAL is not enabled, perform a direct export
+				if errExecute := prwe.export(ctx, requests); errExecute != nil {
+					// We will log the error here, but not return it immediately.
+					// Instead, we'll let the worker complete its execution.
+					prwe.settings.Logger.Error("Failed to export data:", zap.Error(errExecute))
+				}
+			} else {
+				// WAL is enabled, persist requests to the WAL for later export
+				if err := prwe.wal.persistToWAL(requests); err != nil {
+					// Log the error and return a permanent consumer error
+					err = consumererror.NewPermanent(err)
+					prwe.settings.Logger.Error("Failed to persist to WAL:", zap.Error(err))
+				}
+			}
+		}(partitions[i])
 	}
+
+	// Wait for all workers to finish processing the requests
+	wg.Wait()
+
 	return nil
 }
 
-// export sends a Snappy-compressed WriteRequest containing TimeSeries to a remote write endpoint in order
-func (prwe *prwExporter) export(ctx context.Context, requests []*prompb.WriteRequest) error {
-	input := make(chan *prompb.WriteRequest, len(requests))
-	for _, request := range requests {
-		input <- request
+// partitionTimeSeries partitions the time series map into N arrays using hashing/fnv.
+func partitionTimeSeries(tsMap map[string]*prompb.TimeSeries, concurrencyLimit int) []map[string]*prompb.TimeSeries {
```

Comment:
I was reading the code and I checked that the key in the tsMaps is the signature of the metric. The whole point of the tsMap is that each key uniquely identify a time series. Therefore we don't need to calculate the signature here again, we can just trust that the key in the tsMap is a `string` key that uniquely identify the time series.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/translator/prometheusremotewrite/helper.go#L88

As long we always receive a map in which the key is a string, our assumptions will work.